File: app/src/main/java/com/newchar/woolhelper/state/ActivityAttach.java
package com.newchar.woolhelper.state;

import android.app.Activity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.IdRes;
import androidx.appcompat.widget.LinearLayoutCompat;

/**
 * @author wenliqiang@100tal.com
 * date            2019-12-21
 * @since 当前版本描述，
 * @since 迭代版本描述
 */
public class ActivityAttach extends BaseStateLayout<Activity> {

    private ViewGroup mUserSetContainer;

    @IdRes
    private final int activityRootId = android.R.id.content;

    /**
     * 如果你要更改 父布局，
     *
     * @param replaceContainer 切换 状态View 的容器
     */
    public void replaceContainer(ViewGroup replaceContainer) {
        mUserSetContainer = replaceContainer;
    }

    public void switchState(int state) {
        StateLayout stateLayout = layoutArrays.get(state);
        if (stateLayout == null) {
            return;
        }

        ViewGroup rootView;
        if (mUserSetContainer == null) {
            ViewGroup rootViewContainer = host.findViewById(activityRootId);
            if (rootViewContainer.getChildAt(0) instanceof ViewGroup) {
                rootView = (ViewGroup) rootViewContainer.getChildAt(0);
            } else {
                rootView = rootViewContainer;
            }
        } else {
            rootView = mUserSetContainer;
        }
        if (rootView == null) {
            return;
        }
        //把这个View 缓存起来。 从父 View 剔除
        View s = LayoutInflater.from(host).inflate(stateLayout.getLayoutId(), rootView);
        stateLayout.convert(s);
        //---end
        if (rootView instanceof LinearLayout || rootView instanceof LinearLayoutCompat) {
            //如果是线性布局，则加在第一个位置。
            rootView.addView(s, 0, new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
        } else {
            //其他的布局加载最后的位置
            rootView.addView(s);
        }
        this.state = state;
    }




}

File: app/src/main/java/com/newchar/woolhelper/replacer/ActivityAttach.kt
package com.newchar.woolhelper.replacer

import android.R
import android.app.Activity
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.annotation.IdRes
import androidx.appcompat.widget.LinearLayoutCompat

/**
 * @author wenliqiang@100tal.com
 * date            2019-12-21
 * @since 当前版本描述，
 * @since 迭代版本描述
 */
class ActivityAttach(override val host: Activity?) : BaseStateLayout<Activity?>() {
    private var mUserSetContainer: ViewGroup? = null
    @IdRes
    private val activityRootId = R.id.content

    /**
     * 如果你要更改 父布局，
     *
     * @param replaceContainer 切换 状态View 的容器
     */
    fun replaceContainer(replaceContainer: ViewGroup?) {
        mUserSetContainer = replaceContainer
    }

    fun switchState(state: Int) {
        val stateLayout = layoutArrays[state] ?: return
        val rootView: ViewGroup?
        rootView = if (mUserSetContainer == null) {
            val rootViewContainer = host!!.findViewById<ViewGroup>(activityRootId)
            if (rootViewContainer.getChildAt(0) is ViewGroup) {
                rootViewContainer.getChildAt(0) as ViewGroup
            } else {
                rootViewContainer
            }
        } else {
            mUserSetContainer
        }
        if (rootView == null) {
            return
        }
        //把这个View 缓存起来。 从父 View 剔除
        val s =
            LayoutInflater.from(host).inflate(stateLayout.layoutId, rootView)
        stateLayout.convert(s)
        //---end
        if (rootView is LinearLayout || rootView is LinearLayoutCompat) { //如果是线性布局，则加在第一个位置。
            rootView.addView(
                s,
                0,
                ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT
                )
            )
        } else { //其他的布局加载最后的位置
            rootView.addView(s)
        }
        this.state = state
    }

}